include/lemlib/logger/baseSink.hpp

```diff
@@ -0,0 +1,212 @@
+ * logging implementation. A sink could send information to anything, to stdout, to a file, or even the UI on the
+ * brain screen.
+ */
+class BaseSink {
+    public:
+        BaseSink() = default;
+
+        /**
+         * @brief Construct a new combined sink
+         *
+         * <h3> Example Usage </h3>
+         * @code
+         * lemlib::BaseSink combinedSink({lemlib::telemetrySink(), lemlib::infoSink()});
+         * combinedSink.info("This will be sent to both sinks!");
+         * @endcode
+         *
+         * @param sinks The sinks that will have messages sent to them when
+         */
+        BaseSink(std::initializer_list<std::shared_ptr<BaseSink>> sinks);
+
+        /**
+         * @brief Set the lowest level.
+         * If this is a combined sink, this operation will
+         * apply for all the parent sinks.
+         * @param level
+         *
+         * If messages are logged that are below the lowest level, they will be ignored.
+         * The hierarchy of the levels is as follows:
+         * - INFO
+         * - DEBUG
+         * - WARN
+         * - ERROR
+         * - FATAL
+         */
+        void setLowestLevel(Level level);
+
+        /**
+         * @brief Log a message at the given level
+         * If this is a combined sink, this operation will
+         * apply for all the parent sinks.
+         *
+         * @tparam T
+         * @param level The level at which to send the message.
+         * @param format The format that the message will use. Use "{}" as placeholders.
+         * @param args The values that will be substituted into the placeholders in the format.
+         *
+         * <h3> Example Usage </h3>
+         * @code
+         * sink.log(lemlib::Level::INFO, "{} from the logger!", "Hello");
+         * @endcode
+
+         */
+        template <typename... T> void log(Level level, fmt::format_string<T...> format, T&&... args) {
+            if (!sinks.empty()) {
+                for (std::shared_ptr<BaseSink> sink : sinks) { sink->log(level, format, std::forward<T>(args)...); }
+                return;
+            }
+
+            if (level < lowestLevel) { return; }
+
+            // substitute the user's arguments into the format.
+            std::string messageString = fmt::format(format, std::forward<T>(args)...);
+
+            Message message = Message {.level = level, .time = pros::millis()};
+
+            // get the arguments
+            fmt::dynamic_format_arg_store<fmt::format_context> formattingArgs = getExtraFormattingArgs(message);
+
+            formattingArgs.push_back(fmt::arg("time", message.time));
+            formattingArgs.push_back(fmt::arg("level", message.level));
+            formattingArgs.push_back(fmt::arg("message", messageString));
+
+            std::string formattedString = fmt::vformat(logFormat, std::move(formattingArgs));
+
+            message.message = std::move(formattedString);
+
+            sendMessage(std::move(message));
```

Comment:
Remove unnecessary newlines
```suggestion
            std::string formattedString = fmt::vformat(logFormat, std::move(formattingArgs));
            message.message = std::move(formattedString);
            sendMessage(std::move(message));
```